Modules/_sqlite/connection.c

```diff
@@ -440,40 +442,56 @@ remove_callbacks(sqlite3 *db)
 static void
 connection_close(pysqlite_Connection *self)
 {
-    if (self->db) {
-        if (self->autocommit == AUTOCOMMIT_DISABLED &&
-            !sqlite3_get_autocommit(self->db))
-        {
-            /* If close is implicitly called as a result of interpreter
-             * tear-down, we must not call back into Python. */
-            if (_Py_IsInterpreterFinalizing(PyInterpreterState_Get())) {
-                remove_callbacks(self->db);
-            }
-            (void)connection_exec_stmt(self, "ROLLBACK");
+    assert(self->db);
+    if (self->autocommit == AUTOCOMMIT_DISABLED &&
+        !sqlite3_get_autocommit(self->db))
+    {
+        /* If close is implicitly called as a result of interpreter
+         * tear-down, we must not call back into Python. */
+        if (_Py_IsInterpreterFinalizing(PyInterpreterState_Get())) {
+            remove_callbacks(self->db);
         }
+        (void)connection_exec_stmt(self, "ROLLBACK");
+    }
 
-        free_callback_contexts(self);
+    free_callback_contexts(self);
 
-        sqlite3 *db = self->db;
-        self->db = NULL;
+    sqlite3 *db = self->db;
+    self->db = NULL;
 
-        Py_BEGIN_ALLOW_THREADS
-        int rc = sqlite3_close_v2(db);
-        assert(rc == SQLITE_OK), (void)rc;
-        Py_END_ALLOW_THREADS
+    Py_BEGIN_ALLOW_THREADS
+    int rc = sqlite3_close_v2(db);
+    assert(rc == SQLITE_OK), (void)rc;
+    Py_END_ALLOW_THREADS
+}
+
+static void
+connection_finalize(PyObject *self)
+{
+    pysqlite_Connection *con = (pysqlite_Connection *)self;
+    PyObject *exc = PyErr_GetRaisedException();
+    /* Clean up if user has not called .close() explicitly. */
+    if (con->db) {
+        if (PyErr_ResourceWarning(self, 1, "unclosed %R", self)) {
+            /* Spurious errors can appear at shutdown */
+            if (PyErr_ExceptionMatches(PyExc_Warning)) {
+                PyErr_WriteUnraisable(self);
```

Comment:
This log doesn't give much context about why an exception is logged here. I created issue #108082 to discuss this problem.

Reply:
For the user, the current warning ends up looking like this:

> ResourceWarning: unclosed <sqlite3.Connection object at 0x1052589e0>

I guess we could use a more informative warning message. I'll think of something.

Reply:
Oh ok, maybe it's enough.

Reply:
Perhaps _"unclosed database in <sqlite3.Connection object at ...>"_

Reply:
I applied my suggestion in 980d4d2a446db058434bce529fe77efde59618f3